crucible-mir/src/Mir/Trans.hs

```diff
@@ -1375,6 +1382,76 @@ evalRval rv@(M.CopyForDeref lv) = evalLvalue lv
 evalRval rv@(M.ShallowInitBox op ty) = mirFail
     "evalRval: ShallowInitBox not supported"
 
+-- | Construct a `MirExp` representing a union, and initialize it with a
+-- particular field, specified by a field index and an expression. The
+-- expression's type must match the field's type in the provided `M.Adt`.
+buildUnion ::
+  M.Adt ->
+  -- | The index of the field being used to initialize the union
+  Int ->
+  MirExp s ->
+  MirGenerator h s ret (MirExp s)
+buildUnion unionAdt fieldIdx (MirExp fieldTpr fieldExpr) = do
+  unionFields <- case unionAdt ^. M.adtvariants of
+    [v] -> pure (v ^. M.vfields)
+    [] -> die "no variants?"
+    _ -> die "multiple variants?"
+
+  unionField <- case unionFields ^? ix fieldIdx of
+    Just field -> pure field
+    Nothing -> die $ "field index " <> show fieldIdx <> " out of range"
+
+  Some expectedFieldTpr <- tyToReprM (unionField ^. M.fty)
+  case testEquality expectedFieldTpr fieldTpr of
+    Just _ -> pure ()
+    Nothing -> die $
+      "expected field to have type " <> show expectedFieldTpr <>
+      ", but it was " <> show fieldTpr
+
+  emptyAg <- mirAggregate_uninit unionSize
+  fullAg <- mirAggregate_set fieldOffset fieldSize fieldTpr fieldExpr emptyAg
+  pure (MirExp MirAggregateRepr fullAg)
+  where
+    unionSize = fromIntegral (unionAdt ^. M.adtSize)
```

Comment:
I think this needs to be 1 for now.  Otherwise a union with all zero-sized fields will have total size 0 but will have a field covering bytes 0..1 (even ZSTs are treated as size 1 for now), which is illegal.  Might be worth adding a test for this (initialize an instance of `union U { x: (), y: () }`).

Reply:
I changed this to size 1 in 4612fe406b99cb2a11258199828875e39f447cd1. However, I wasn't able to add a test case involving initializing a ZST union that resulted in the union actually being evinced in MIR.